django/utils/functional.py

```diff
@@ -171,26 +151,54 @@ def __ge__(self, other):
         def __hash__(self):
             return hash(self.__cast())
 
-        def __mod__(self, rhs):
-            return self.__cast() % rhs
+        def __format__(self, format_spec):
+            return format(self.__cast(), format_spec)
+
+        # Explicitly wrap methods which are required for certain operations on
+        # int/str objects to function correctly.
+
+        def __mod__(self, other):
+            return self.__cast() % other
+
+        def __rmod__(self, other):
+            return other % self.__cast()
 
         def __add__(self, other):
             return self.__cast() + other
 
         def __radd__(self, other):
             return other + self.__cast()
 
-        def __deepcopy__(self, memo):
-            # Instances of this class are effectively immutable. It's just a
-            # collection of functions. So we don't need to do anything
-            # complicated for copying.
-            memo[id(self)] = self
-            return self
+        def __mul__(self, other):
+            return self.__cast() * other
+
+        def __rmul__(self, other):
+            return other * self.__cast()
```

Comment:
Do we need `__rmul__` or `__rmod__`? :thinking: I don't any test failures after removing them.

Reply:
Ahhh... I wasn't really sure why `__radd__` was necessary, but seems to be required for the `str` case...

This explains why:

```pycon
>>> for type in (int, str):
...     for method in ("__add__", "__radd__", "__mul__", "__rmul__", "__mod__", "__rmod__"):
...         print(f"Has {type.__name__}.{method}? {hasattr(type, method)}")
... 
Has int.__add__? True
Has int.__radd__? True
Has int.__mul__? True
Has int.__rmul__? True
Has int.__mod__? True
Has int.__rmod__? True
Has str.__add__? True
Has str.__radd__? False
Has str.__mul__? True
Has str.__rmul__? True
Has str.__mod__? True
Has str.__rmod__? True

```

Reply:
Will remove the ones that we don't require.